architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,302 @@
+    * 4-bit quantization kernels to reduce memory storage of the base weights.
+- [Unsloth](https://github.com/unslothai/unsloth).
+    * Fused operation kernels
+    * Kernels for common model architectures (e.g., cross-entropy losses, RoPE embeddings and RMS norms).
+- [megablocks](https://github.com/databricks/megablocks).
+    * acceleration package for distributing mixture-of-experts that improves upon FSDP sharding.
+
+<!--
+Why this is a valuable problem to solve? What background information is needed to show how this design addresses the problem?
+
+Which users are affected by the problem? Why is it a problem? What data supports this? What related work exists?
+-->
+
+### User Benefit
+
+Users will benefit from powerful training tools integrated into the platform, that are not readily accessible from huggingface. With these tools, users will be able to train models with less GPU resources and/or quicker, resulting in quicker turnaround and improved user experience. 
+
+<!--
+How will users (or other contributors) benefit from this work? What would be the headline in the release notes or blog post?
+-->
+
+## Decision
+
+Terminology | Description
+--|--
+Maintainer | Developers of `sft_trainer.py`.
+Framework | an extensible `Framework` managing all implemented methods.
+Framework Plugin | Self-contained implementation of a framework method.
+
+The proposal satisfies the following desiredata:
+- Unified configuration YAML for all plugins. Fine configuration details abstracted away from Maintainers and other plugin developers.
+- Modular design allows new methods plugins to be added / removed / deactivated seemlessly.
+- Modular design enforces that plugins interact with `sft_trainer.py` at controlled points, and throw appropriate exceptions.
+- Generic enough for most use cases of interest (e.g., quantization, distributed training, etc).
+- Unobstrusive design that only *modifies the model*, and leaves [`SFTTrainer`] unmodified. Minimal inversion-of-control maintained through [`TrainerCallbacks`].
+
+### Only the Model is Modified
+
+Since the [`SFTTrainer`] is designed to work with generic pytorch models, modifying the model is much less intrusive to the training pipeline, then say, modifying [`SFTTrainer`] itself. The hope is then if we constrain ourselves to modify only the model, that we can implement all the method plugins (e.g., quantization, distributed training, etc) that we hope for. 
+
+The framework is designed to only modify them model at two integration points in `sft_trainer.py`. The primary motivation for this is easy code maintenance:
+1. an *optional* `loading` method that acts as a drop-in replacement for `AutoModel.from_pretrained`. 
+2. an *optional* `agumentation` method that provides a way to perform *minor* adjustments to an already instantiated model 
+3. an *optional* `callback` method to install `TrainerCallbacks` (if needed, e.g. custom save logic).
+
+```python
+class TuningAccelerationPlugin:
```

Comment:
Maybe it would be helpful to have an example, e.g., what would an implementation of this look like if Lora or PrefixTuning were implemented using this plugin architecture?

If I'm understanding correctly, I'd expect that `augmentation` would be used to do things like replacing Linear layers with LoraLinear layers, and `loading` would be used to load a Lora checkpoint (but why 'loading' instead of just 'load'?). I'm not sure where `callbacks` would be applied, and not sure how the checkpoint saving code would be modified.

Reply:
Yes your comments on `augmentation` and `loading` are spot on. Yes we could consider to change to `load` (however in the revision I changed to `model_loader`, but that could be further changed.

An example of custom model saving logic could be something like as follows. Back in the days when PEFT was new, the HFTrainer would not know how to detect that it is a PEFT-Model, and save the full model weights. So 
we used to have some logic to delete and replace with the adapter weights only. 

In the examples I have now, this is not the case, but I included it in the design as it could be something that is needed in the future. 

```
class SavePeftModelCallback(TrainerCallback):
    def on_save(
        self,
        args: TrainingArguments,
        state: TrainerState,
        control: TrainerControl,
        **kwargs,
    ):
        checkpoint_folder = os.path.join(args.output_dir, f"{PREFIX_CHECKPOINT_DIR}-{state.global_step}")

        peft_model_path = os.path.join(checkpoint_folder, "adapter_model")
        kwargs["model"].save_pretrained(peft_model_path)

        pytorch_model_path = os.path.join(checkpoint_folder, "pytorch_model.bin")
        if os.path.exists(pytorch_model_path):
            os.remove(pytorch_model_path)
        return control
```

Reply:
hmm...  Currently PEFT is handled by PEFT-specific code inside SFTTrainer (https://github.com/huggingface/trl/blob/v0.8.6/trl/trainer/sft_trainer.py), including code in the superclass (https://github.com/huggingface/transformers/blob/v4.40.1/src/transformers/trainer.py#L2500) that's specific to how adapters are checkpointed. Is none of that code needed and something like this can be implemented entirely external to the trainer class?

If I go look at the docs for callbacks (https://huggingface.co/docs/transformers/en/main_classes/callback) it says `Callbacks are “read only” pieces of code`, but I guess "read only" only means from the perspective of modifying the behavior of the training loop, so I guess writing checkpoints is still within the idea of "read only?"

Reply:
> If I go look at the docs for callbacks (https://huggingface.co/docs/transformers/en/main_classes/callback) it says Callbacks are “read only” pieces of code, but I guess "read only" only means from the perspective of modifying the behavior of the training loop, so I guess writing checkpoints is still within the idea of "read only?"

Correct, I believe because "read-only" is in quotes, it is not to be taken literally. For example
- `class transformers.integrations.TensorBoardCallback` may write to some location to populate data for tensorboard
- [`PrinterCallback`](https://github.com/huggingface/transformers/blob/9fe3f585bb4ea29f209dc705d269fbe292e1128f/src/transformers/trainer_callback.py#L220-L224)  modifies the `log` object.

And the fact that `TrainerCallbacks` [provide an `on_save` event](https://github.com/huggingface/transformers/blob/9fe3f585bb4ea29f209dc705d269fbe292e1128f/src/transformers/trainer_callback.py#L289-L293) suggests it is meant to allow custom save logic to be implemented, hence the example above.

> hmm... Currently PEFT is handled by PEFT-specific code inside SFTTrainer (https://github.com/huggingface/trl/blob/v0.8.6/trl/trainer/sft_trainer.py), including code in the superclass (https://github.com/huggingface/transformers/blob/v4.40.1/src/transformers/trainer.py#L2500) that's specific to how adapters are checkpointed. Is none of that code needed and something like this can be implemented entirely external to the trainer class?

If you move down a couple of lines you would see [that this loading logic works only if the adapters are saved properly](https://github.com/llohann-speranca/transformers/blob/5a35cb9c12724f97b3866edd260cf1f27613f026/src/transformers/trainer.py#L2077-L2081), and suggest that in order to save properly, we should use a [custom callback](https://github.com/huggingface/peft/issues/96). The loading logic was fixed around mid 2023, and only on Jan this year, that they introduced a [`save_adapter` into `save_fsdp_model` in accelerate](https://github.com/huggingface/transformers/pull/28297). So from Jan onwards, this callback is no more needed, but within those 6 months, you would have not been able to resume your trainings without a proper callback.

This is a situtation that no longer exists, but within the context of `tuning.sft_trainer.py`, I believe we need an organized way of putting in such fixes without always changing the `sft_trainer.py`. Hence this Framework is envisioned to do that. One year ago PEFT was envisioned as the method to train cheaply, but it wouldnt have worked well without having properly saved checkpoints.